app/models/sendgrid_event.rb

```diff
@@ -0,0 +1,74 @@
+# frozen_string_literal: true
+class SendgridEvent < ApplicationRecord
+
+  # TODO: Add test coverage for SendgridEvent#email
+  store_accessor :payload, :email
+
+  # TODO: Test coverage
+  # To make allowances for occasional inbox down time, this counts a maximum of one fail per day,
+  # e.g.:
+  #
+  # - If 2 emails are sent on the same day, and both fail, then this counts as only 1 fail, not
+  #   2 fails.
+  #
+  # - If 2 emails are sent, one each on 2 consecutive days, and both fail, then this counts as 2
+  #   fails, as the failures were on different days.
+  def self.fails_since_last_delivery(email)
+    # TODO: Test if last_delivered_at can ever be NULL/nil, set as zero if so.
+    last_delivered_at = where_email(email).where_event_type("delivered").maximum(timestamp_sql)
+
+    # TODO: Test works if last_delivered_at is NULL/nil (assuming this can happen)
+    fails_since_last_delivery =
+      select("DISTINCT(date_trunc('day', #{timestamp_sql}))")
+        .where_email(email)
+        .where_event_type("dropped", "bounce")
+        .where("#{timestamp_sql} > ?", last_delivered_at)
+        .count
+  end
+
+  def self.where_email(email)
+    where("(payload->>'email') = ?", email)
+  end
+
+  def self.where_event_type(*event_types)
+    where("(payload->>'event') IN (?)", event_types)
+  end
+
+  def self.where_sg_event_id(sg_event_id)
+    where("(payload->>'sg_event_id') = ?", sg_event_id)
+  end
+
+  def self.timestamp_sql
+    "to_timestamp((payload->>'timestamp')::double precision)"
+  end
+
+  def self.process_later(payload)
+    # TODO: Test coverage duplicates are prevented
+    return if where_sg_event_id(payload[:sg_event_id]).exists?
+
+    transaction do
+      event = create!(payload: payload, status: "pending")
```

Comment:
The main reason is to preserve the original payload. We may discover that other data in the payload becomes original. Also storing the data in additional columns feels like additional work that there isn't a need for yet. Can be added in a later PR if it becomes useful.

Reply:
>  I don't have access to sendgrid but I am guessing their UI has this info.

IIRC SendGrid discard events, it is not available in their UI, or if it is, only in the short-term.

Reply:
I think it will be better to store needed columns for queries as a separated columns and we can keep also the rest of the paylod in JSONB non-indexed column. It will simplify the code since those JSONB scopes will not be needed.

Reply:
yeah, `payload->>'timestamp'` looks ugly. Can we please store the column we need?

Reply:
Understood, dedicated columns are preferred over JSONB, will change as requested 👍

Reply:
Latest commit includes dedicated columns.